File: Clients/NET/MessageConversions.cs
using Google.Protobuf.WellKnownTypes;
using System.Numerics;
using GrpcTuple = LindaSharp.Services.Tuple;
using GrpcPattern = LindaSharp.Services.Pattern;
using GrpcScriptExecutionStatus = LindaSharp.Services.ScriptExecutionStatus;
using GrpcScript = LindaSharp.Services.Script;
using System.Collections;
using static LindaSharp.IScriptEvalLinda;
using static LindaSharp.IScriptEvalLinda.ScriptExecutionStatus;

namespace LindaSharp.Client;

internal static class MessageConversions {
	internal static Value ElemToValue(object? elem) {
		return elem switch {
			null => Value.ForNull(),
			bool b => Value.ForBool(b),
			string str => Value.ForString(str),

			// TODO: Simplify
			sbyte num => Value.ForNumber(num),
			byte num => Value.ForNumber(num),
			short num => Value.ForNumber(num),
			ushort num => Value.ForNumber(num),
			int num => Value.ForNumber(num),
			uint num => Value.ForNumber(num),
			long num => Value.ForNumber(num),
			ulong num => Value.ForNumber(num),
			nint num => Value.ForNumber(num),
			nuint num => Value.ForNumber(num),
			float num => Value.ForNumber(num),
			double num => Value.ForNumber(num),
			decimal num => Value.ForNumber((double)num),
			BigInteger num => Value.ForNumber((double)num),

			IDictionary dict => Value.ForStruct(dict.ToStruct()),
			IEnumerable list => Value.ForList(list.Cast<object>().Select(ElemToValue).ToArray()),
			_ => throw new ArgumentException($"Cannot parse {elem}")
		};
	}

	private static object? ValueToElem(Value value) {
		return value.KindCase switch {
			Value.KindOneofCase.None => throw new NotImplementedException(),
			Value.KindOneofCase.NullValue => null,
			Value.KindOneofCase.NumberValue => value.NumberValue,
			Value.KindOneofCase.StringValue => value.StringValue,
			Value.KindOneofCase.BoolValue => value.BoolValue,
			Value.KindOneofCase.StructValue => value.StructValue.Fields.ToDictionary(p => p.Key, p => ValueToElem(p.Value))!,
			Value.KindOneofCase.ListValue => value.ListValue.Values.Select(ValueToElem).ToList(),
			_ => throw new ArgumentException($"Unknown case {value.KindCase}")
		};
	}

	private static Struct ToStruct(this IDictionary dict) {
		var s = new Struct();

		foreach (DictionaryEntry entry in dict)
			s.Fields.Add(entry.Key.ToString(), ElemToValue(entry.Value));

		return s;
	}

	public static GrpcTuple ToGrpcTuple(this object[] lindaTuple) {
		var grpcTuple = new GrpcTuple();
		grpcTuple.Fields.AddRange(lindaTuple.Select(ElemToValue));

		return grpcTuple;
	}

	public static GrpcPattern ToGrpcPattern(this object?[] lindaPattern) {
		var grpcPattern = new GrpcPattern();
		grpcPattern.Fields.AddRange(lindaPattern.Select(ElemToValue));

		return grpcPattern;
	}

	public static object[] ToLindaTuple(this GrpcTuple grpcTuple) => grpcTuple.Fields.Select(ValueToElem).ToArray()!;

	public static ScriptExecutionStatus ToLindaStatus(this GrpcScriptExecutionStatus status) {
		return status.StatusCase switch {
			GrpcScriptExecutionStatus.StatusOneofCase.NotFound => new ScriptExecutionStatus(ExecutionState.NotFound),
			GrpcScriptExecutionStatus.StatusOneofCase.Ok => new ScriptExecutionStatus(ExecutionState.Finished),
			GrpcScriptExecutionStatus.StatusOneofCase.Exception => new ScriptExecutionStatus(
				ExecutionState.Exception,
				new Exception(status.Exception.Message) {
					Source = status.Exception.Source,
					// TODO: Embed other fields
				}),
			_ => throw new ArgumentException(nameof(status.StatusCase))
		};
	}

	public static GrpcScript ToGrpcScript(this Script script) {
		return new GrpcScript {
			Code = script.Code,
			Type = script.Type switch {
				Script.Language.IronPython => GrpcScript.Types.Type.Ironpython,
				Script.Language.CSharp => GrpcScript.Types.Type.CSharp,
				_ => throw new ArgumentException(nameof(script.Type))
			}
		};
	}
}

File: Clients/Python/message_conversions.py
from google.protobuf.internal.well_known_types import ListValue, Struct
from google.protobuf.struct_pb2 import Value
from linda import IScriptEvalLinda
from scripts_pb2 import Script as GrpcScript
from health_pb2 import ScriptExecutionStatus as GrpcScriptExecutionStatus

def elem_to_value(elem):
	match elem:
		case bool():
			return Value(bool_value=elem)
		case str():
			return Value(string_value=elem)
		case int():
			return Value(number_value=elem)
		case float():
			return Value(number_value=elem)
		case list():
			grpc_list = ListValue()
			grpc_list.values.extend(elem_to_value(e) for e in elem)
			return Value(list_value=grpc_list)
		case dict():
			grpc_struct = Struct()
			grpc_struct.update({k: elem_to_value(v) for k, v in elem.items()})
			return Value(struct_value=grpc_struct)
		case None:
			return Value(null_value=0)
		case _:
			raise ValueError(f"Unsupported type {type(elem)}")


def value_to_elem(value: Value):
	if value.HasField("bool_value"):
		return value.bool_value
	elif value.HasField("string_value"):
		return value.string_value
	elif value.HasField("number_value"):
		return value.number_value
	elif value.HasField("list_value"):
		return [value_to_elem(v) for v in value.list_value.values]
	elif value.HasField("struct_value"):
		return {k: value_to_elem(v) for k, v in value.struct_value.fields.items()}
	elif value.HasField("null_value"):
		return None
	else:
		raise ValueError(f"Unsupported type {value}")


def to_grpc_script(script: IScriptEvalLinda.Script) -> GrpcScript:
	match script:
		case IScriptEvalLinda.Script(IScriptEvalLinda.Script.Type.IRONPYTHON, code):
			return GrpcScript(type=GrpcScript.Type.IRONPYTHON, code=code)
		case IScriptEvalLinda.Script(IScriptEvalLinda.Script.Type.C_SHARP, code):
			return GrpcScript(type=GrpcScript.Type.C_SHARP, code=code)
		case _:
			raise ValueError(f"Unsupported script type {script.type}")

def from_grpc_script_execution_status(status: GrpcScriptExecutionStatus) -> GrpcScriptExecutionStatus:
	if status.HasField("not_found"):
		return IScriptEvalLinda.ScriptExecutionStatus.NOT_FOUND
	elif status.HasField("ok"):
		return IScriptEvalLinda.ScriptExecutionStatus.FINISHED
	elif status.HasField("exception"):
		grpc_extension = status.exception
		#TODO: Set exception message
		return IScriptEvalLinda.ScriptExecutionStatus.EXCEPTION